modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java

```diff
@@ -0,0 +1,449 @@
+            }
+        }
+
+        logger.trace("Updating geoip databases");
+        IngestGeoIpMetadata geoIpMeta = clusterState.metadata().custom(IngestGeoIpMetadata.TYPE, IngestGeoIpMetadata.EMPTY);
+
+        // if there are entries in the cs that aren't in the persistent task state,
+        // then download those (only)
+        // ---
+        // if there are in the persistent task state, that aren't in the cluster state
+        // then nuke those (only)
+        // ---
+        // else, just download everything
+        boolean addedSomething = false;
+        {
+            EnterpriseGeoIpTaskState _state = state;
+            Set<String> metas = Set.copyOf(_state.getDatabases().keySet());
+            for (Map.Entry<String, DatabaseConfigurationMetadata> entry : geoIpMeta.getDatabases().entrySet()) {
+
+                final String id = entry.getKey();
+                DatabaseConfiguration database = entry.getValue().database();
+                if (metas.contains(database.name() + ".mmdb") == false) {
+                    logger.debug("A new database appeared [{}]", database.name());
+
+                    final String accountId = database.maxmind().accountId();
+                    try (HttpClient.PasswordAuthenticationHolder holder = credentialsBuilder.apply(accountId)) {
+                        if (holder == null) {
+                            logger.warn("No credentials found to download database [{}], skipping download...", id);
+                        } else {
+                            processDatabase(holder.get(), id, database);
+                        }
+                    }
+
+                    addedSomething = true;
+                }
+            }
+        }
+
+        boolean droppedSomething = false;
+        {
+            // rip anything out of the task state that doesn't match what's in the cluster state,
+            // that is, if there's no longer an entry for a database in the repository,
+            // then drop it from the task state, too
+            Set<String> databases = geoIpMeta.getDatabases()
+                .values()
+                .stream()
+                .map(c -> c.database().name() + ".mmdb")
+                .collect(Collectors.toSet());
+            EnterpriseGeoIpTaskState _state = state;
+            Collection<Map.Entry<String, Metadata>> metas = List.copyOf(_state.getDatabases().entrySet());
+            for (Map.Entry<String, Metadata> entry : metas) {
+                String name = entry.getKey();
+                Metadata meta = entry.getValue();
+                if (databases.contains(name) == false) {
+                    logger.debug("Dropping [{}], databases was {}", name, databases);
+                    _state = _state.remove(name);
+                    deleteOldChunks(name, meta.lastChunk() + 1);
+                    droppedSomething = true;
+                }
+            }
+            state = _state;
+            updateTaskState();
```

Comment:
Should this unconditionally update the state, or should this be inside a `if (droppedSomething) { updateTaskState(); }` block?

Reply:
I think you are right (and I committed that change). @joegallo can chime in if there's something to the logic we're missing.

Reply:
Nah, I think I'd intended it to be what you ended up doing, I just didn't see the miss. Thanks for catching it!